sd_bandits/obp_extensions/policy.py

```diff
@@ -85,3 +88,108 @@ def update_params(self, action: int, reward: float) -> None:
+        Name of bandit policy.
+        
+    """
+    
+    base_policy: BaseContextFreePolicy
+    n_segments: int
+    policy_name: str = ""
+    
+    def __post_init__(self) -> None:
+        """Initialize Class."""
+        assert isinstance(
+            self.base_policy, BaseContextFreePolicy
+        ), "Supplied policy must be an instance of BaseContextFreePolicy"
+        assert isinstance(self.n_segments, int), f"n_segments must be an integer, you supplied a {type(self.n_segments)}"
+        
+        # then instantiate all of base policies attributes
+        self.n_actions = self.base_policy.n_actions
+        self.len_list = self.base_policy.len_list
+        self.batch_size = self.base_policy.batch_size
+        self.random_state = self.base_policy.random_state
+        self.policy_type = "segmented"
+        self.n_trial = 0
+        
+        # give name if no name given
+        if self.policy_name == "":
+            self.policy_name = f'{self.base_policy.policy_name}_seg'
+        
+        # create dictionary of policy objects
+        segment_policy_pairs = [(i, deepcopy(self.base_policy)) for i in range(self.n_segments)]
+        self.segment_policies = dict(segment_policy_pairs)
+        
+    @property
+    def action_counts(self):
+        """Returns action counts"""
+        action_counts = np.zeros(self.n_actions)
+        for i in range(self.n_segments):
+            action_counts += self.segment_policies[i].action_counts
+        return action_counts
+    
+    @property
+    def reward_counts(self):
+        """Returns reward counts"""
+        reward_counts = np.zeros(self.n_actions)
+        for i in range(self.n_segments):
+            reward_counts += self.segment_policies[i].action_counts * self.segment_policies[i].reward_counts
+        reward_counts = np.divide(reward_counts, self.action_counts, where=self.action_counts != 0)
+        return reward_counts
+        
+        
+    def select_action(self, segment: int) -> np.ndarray:
+        """Select a list of actions.
+        
+        Parameters
+        segment: int
+            The user's segment
+        
+        Returns
+        ----------
+        selected_actions: array-like, shape (len_list, )
+            List of selected actions.
+
+        """
+        
+        return self.segment_policies[segment].select_action()
+
+
+    def update_params(self, action: int, reward: float, segment: int) -> None:
+        """Update policy parameters.
+        Same as obp.policy.EpsilonGreedy
+
+        Parameters
+        ----------
+        action: int
+            Selected action by the policy.
+
+        reward: float
+            Observed reward for the chosen action and position.
+        
+        segment: int
+            The user's segment
+
+        """
+        # first update n_trial for the policy
+        self.n_trial += 1
+        self.segment_policies[segment].n_trial = self.n_trial - 1
```

Comment:
I don't fully understand this line. Why not just `self.segment_policies[segment].n_trial += 1`?

Reply:
We want to update every 'batch_size' trials overall, but each segment's policy object has a different n_trials value since each segment's policy may have been accessed a different number of times. To get them all to update every 'batch_size' trials, we have to make sure they have the same n_trials value. And I do `self.n_trial - 1` because when you call update_params, it immediately adds 1 to n_trial.
However, I do have to rethink this a little because it only updates when `n_trial % batch_size == 0`